conn.go

```diff
@@ -304,14 +304,14 @@ func (c *conn) Start() error {
 
 	// run connection establishment state machine
 	for state := c.negotiateProto; state != nil; {
-		state = state()
-	}
-
-	// check if err occurred
-	if c.err != nil {
-		close(c.txDone) // close here since connWriter hasn't been started yet
-		_ = c.Close()
-		return c.err
+		var err error
+		state, err = state()
```

Comment:
Should we be setting c.err here now? It's not in the sasl state functions anymore and the connection is in a failed state.

Reply:
There's no need to set `c.err` if this fails.  If `Start()` returns an error no `*conn`, and thus no `*Client` is returned.

Reply:
Looking at it more, it seems right. There's a bit of inconsistency but it's an internal struct. TBH, I'd much prefer if you can't even get a 'conn' instance without the negotiation being done, similar to .Dial().